pkg/typesystem/typesystem.go

```diff
@@ -379,6 +387,59 @@ func (t *TypeSystem) UsersetCanFastPath(relationReferences []*openfgav1.Relation
 	return true
 }
 
+// recursiveUsersetNodeCanFastpath is a helper function to determine whether the node (object#relation) has
+// - one edge back to itself
+// - other edges lead directly to terminal types (not union/intersection/exclusion).
+func (t *TypeSystem) recursiveUsersetNodeCanFastpath(curAuthorizationModelNode *graph.AuthorizationModelNode, userType string) bool {
+	hasNodeBackToItself := false
+	hasCorrectTerminalType := false
+	childrenNodes := t.authorizationModelGraph.To(curAuthorizationModelNode.ID())
+	for childrenNodes.Next() {
+		curNode := childrenNodes.Node()
+		curChildAuthorizationModelNode, ok := curNode.(*graph.AuthorizationModelNode)
+		if !ok {
+			// Strange, I would have expected to be able to cast to AuthorizationModelNode
+			return false
+		}
+		childNodeObjectType, childNodeRelation := tuple.SplitObjectRelation(curChildAuthorizationModelNode.Label())
+		if childNodeRelation == "" {
+			// this means the childNode is either a terminal type or one of relation keyword
+			if childNodeObjectType == userType {
+				hasCorrectTerminalType = true
+			} else if childNodeObjectType == "union" || childNodeObjectType == "intersection" || childNodeObjectType == "exclusion" {
+				// they are not terminal type and are not allowed
+				return false
+			}
+		} else {
+			if curAuthorizationModelNode == curChildAuthorizationModelNode {
+				hasNodeBackToItself = true
+			} else {
+				return false
+			}
+		}
+	}
+	return hasNodeBackToItself && hasCorrectTerminalType
+}
+
+// RecursiveUsersetCanFastPath returns whether the specified object type and relation allows
+// for optimization.
+func (t *TypeSystem) RecursiveUsersetCanFastPath(objectTypeRelation string, userType string) bool {
+	nodes := t.authorizationModelGraph.Nodes()
```

Comment:
Is there a way this could be faster? We are basically iterating through a list every time we ask this. We could memoize it or perhaps pre compute this. 

Doesn't need to be addressed in this PR, but this would get linearly worst for all the nodes that are added to the graph, which we add for not just types.

Reply:
Let's compute this at TypeSystem creation time.  This will allow us to avoid casting into *graph.AuthorizationModelNode a few lines later which does not help with performance.

Reply:
This belongs in the graph itself :D

Reply:
See https://github.com/openfga/language/pull/354